designs/nodeclaim-limits.md

```diff
@@ -0,0 +1,44 @@
+# NodeClaim Limits
+
+**NodeClaim**: Karpenter’s representation of a Node request. A NodeClaim declaratively describes how the node should look, including instance type requirements that constrain the options that the Cloud Provider can launch with. For the AWS Cloud Provider, the instance requirements of the NodeClaim translate to the requirements of the CreateFleet request. More details can be found [here](https://github.com/aws/karpenter-provider-aws/blob/main/designs/node-ownership.md).
+
+WIP PR: https://github.com/kubernetes-sigs/karpenter/pull/1151
+
+## Motivation
+
+The Karpenter NodePool Spec allows for specifying [provisioning limits](https://karpenter.sh/docs/concepts/nodepools/#speclimits) which can be used to limit based on any resource type provided by the cloudprovider or device plugins.  If a limit is exceeded, node provisioning is prevented until more resources become available (typically in the form of node terminations).
+
+This works for users who are purely concerned about limiting resource usage within a NodePool, but doesn't work for specifying a global limit across all NodePools or for limiting a higher level representation of usage such as the number of nodes in a cluster (without also reducing flexibility of instance types).
+
+## Use-Cases
+1. I'm using a shared network with limited IP pools and need to ensure Karpenter's usage doesn't consume too many IPs. Each Node consumes a similar number of IPs but low-level resource usage for things like CPU/Memory differs between instance types.
+2. I have a special NodePool (for example a set of high performance nodes or AWS Reserved Instances) with a finite amount of compute available. I need a way to prevent Karpenter from launching too many nodes of this type. The only to do this is by calculating against an available resource like number of CPU's and adding this as a limit. This doesn't target the actual limiter which is number of instances rather than resource consumption.
+3. My CNI can't support more than 400 nodes but I'm not concerned about underlying resource usage. I have no way to restrict Karpenter to this limit without severly limiting instance type options.
+
+Currently, Karpenter users face a limitation in controlling the maximum number of nodes provisioned by Karpenter. This constraint hinders operations in environments where the CNI in use can't scale beyond a certain number of nodes or in scenarios involving shared networks with a finite IP address allocation. Some additional use cases and discussion on this limitation can be seen in the following issues: (#732, aws/karpenter-provider-aws#4462).
+
+## Proposal
+
+### New controller flag `max-node-claims` and per-NodePool limits in the `spec.limits` section of NodePool template
```

Comment:
I'd really prefer to be able to reconfigure this without a controller restart, maybe using a ConfigMap?

What I'm imagining: you specify a NodePool selector, and you specify a cap on NodeClaims owned by the selected NodePool(s). Karpenter doesn't make a NodeClaim if doing so would go over the cap.